wgpu-core/src/device/mod.rs

```diff
@@ -3887,12 +3897,46 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
 
     pub fn device_get_swap_chain_preferred_format<B: GfxBackend>(
         &self,
-        _device_id: id::DeviceId,
-    ) -> Result<TextureFormat, InvalidDevice> {
+        device_id: id::DeviceId,
+        surface_id: id::SurfaceId,
+    ) -> Result<TextureFormat, GetSwapChainPreferredFormatError> {
         span!(_guard, INFO, "Device::get_swap_chain_preferred_format");
-        //TODO: we can query the formats like done in `device_create_swapchain`,
-        // but its not clear which format in the list to return.
-        // For now, return `Bgra8UnormSrgb` that we know is supported everywhere.
+
+        let hub = B::hub(self);
+        let mut token = Token::root();
+
+        let (mut surface_guard, mut token) = self.surfaces.write(&mut token);
+        let (adapter_guard, mut token) = hub.adapters.read(&mut token);
+        let (device_guard, mut _token) = hub.devices.read(&mut token);
+        let device = device_guard
+            .get(device_id)
+            .map_err(|_| DeviceError::Invalid)?;
+        let surface = surface_guard
+            .get_mut(surface_id)
+            .map_err(|_| GetSwapChainPreferredFormatError::InvalidSurface)?;
+
+        let formats = {
+            let surface = B::get_surface_mut(surface);
+            let adapter = &adapter_guard[device.adapter_id.value];
+            let queue_family = &adapter.raw.queue_families[0];
+            if !surface.supports_queue_family(queue_family) {
+                return Err(GetSwapChainPreferredFormatError::UnsupportedQueueFamily);
+            }
+            surface.supported_formats(&adapter.raw.physical_device)
+        };
+        if let Some(formats) = formats {
+            // Prefer Bgra8UnormSrgb
+            if formats.contains(&conv::map_texture_format(TextureFormat::Bgra8UnormSrgb, device.private_features)) {
+                return Ok(TextureFormat::Bgra8UnormSrgb);
+            }
+            // If not available, use Rgba8UnormSrgb
+            if formats.contains(&conv::map_texture_format(TextureFormat::Rgba8UnormSrgb, device.private_features)) {
+                return Ok(TextureFormat::Rgba8UnormSrgb);
+            }
+            // TODO: are there devices that support neither of these?
+        }
+
+        // If no formats were returned, use Bgra8UnormSrgb
         Ok(TextureFormat::Bgra8UnormSrgb)
```

Comment:
we'd also hit this path if the formats are returned, but neither of the branches was taken

Reply:
Do you think it'd be better to return an error in this case? Or just take the first format in the list, even if it's a single-color format?

Reply:
let's take the first format, still better than an error :)

Reply:
I had another look at the definition of `getSwapChainPreferredFormat(adapter)` in the WebGPU spec, where it says:

> Return an optimal GPUTextureFormat to use when creating a GPUSwapChain with the given adapter. Must be one of the supported swap chain formats.
> The supported swap chain formats are a set of GPUTextureFormats that must be supported when specified as a GPUSwapChainDescriptor.format regardless of the given GPUSwapChainDescriptor.device, initially set to: «"bgra8unorm", "bgra8unorm-srgb", "rgba8unorm", "rgba8unorm-srgb"».

So I guess it is sufficient to check for these four, and then fallback to the first format in the list, in the unlikely case that none of these is supported.

Reply:
On the other hand, there are more formats described in HAL than there are texture formats defined in wgpu. So for now, I think checking these four mentioned in the spec and returning an error instead might be the easier choice until someone comes up with a use case where these four don't work.